File: 프로그래머스/1/42889. 실패율/실패율.py
from collections import Counter
def solution(N, stages):
    
    answer = []
    
    stages.sort()
    # print(stages)
    
    counter = Counter(stages)
    
    for stage_number in range(1,N+1):
        # print(counter)
        tmp = [0,stage_number] # [실패율, 스테이지 번호]    
        
        if not counter:
            answer.append(tmp)
            continue
            
        failure =  counter[stage_number] / sum(counter.values())
        tmp[0] = failure
        # print(tmp)
        
        answer.append(tmp)
        del counter[stage_number]
        
    answer.sort(key = lambda x : (-x[0],x[1]) )
    
    return list(map(lambda x: x[1], answer  ) )

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.stream.*;
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        Map<Integer,Double> answer = new HashMap<>();
        
        Map<Integer, Long> stageMap = Arrays.stream(stages)
            .boxed()
            .collect(Collectors.groupingBy(i->i, Collectors.counting()));
        
        // System.out.println(stageMap);
            
        // Arrays.sort(stages);
        int total = stages.length;
        
        for (int i=1; i<= N; i++) {
            long cnt = stageMap.getOrDefault(i, 0L);
            if (cnt == 0) {
                answer.put(i, 0.0);
            }
            else {
                answer.put(i, (double)cnt / total);
                total -= cnt;
            }
        }
        
        return answer.entrySet().stream()
            .sorted(Comparator.comparingDouble(entry-> entry.getValue() * -1))
            .map(entry-> entry.getKey())
            .mapToInt(Integer::intValue)
            .toArray();
    }
}